### PR TITLE
#27500 Prepare Catalog module Tests for PHPUnit 8

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
Preparing PHPUnit Tests to run with PHPUnit 8.
PRs for each module is separated to make tests easier to merge.

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/233

### Fixed Issues (if relevant)
1. magento/magento2#27500

### Manual testing scenarios (*)
N/A

### Questions or comments
PR #27701 needs to be merged first (otherwise PHPUnit tests fail)

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by slavvka** [comment]
Unit tests failure is caused by:
```
18:16:55 PHP Fatal error:  Declaration of Magento\CatalogInventory\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AdvancedInventoryTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/CatalogInventory/Test/Unit/Ui/DataProvider/Product/Form/Modifier/AdvancedInventoryTest.php on line 163
```

File: app/code/Magento/Catalog/Test/Unit/Model/ResourceModel/Product/ImageTest.php [hunk 1]
```diff
@@ -58,7 +56,7 @@ protected function setUp(): void
     /**
      * @return MockObject
      */
-    protected function getVisibleImagesSelectMock(): MockObject
+    protected function getVisibleImagesSelectMock(): \PHPUnit\Framework\MockObject\MockObject
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/ResourceModel/Product/ImageTest.php]: Would be great to import this class, but that's too small thing compared to whole PR, so we can leave it

**Review by ihor-sviziev — APPROVED** [review]
👍Thank you for contribution!

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7446](https://jira.corp.magento.com/browse/ENGCOM-7446) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@slavvka 
I merged some conflicting changes with 2.4-develop branch, that are missing in phpunit8 branch, please merge them.
Also that fatal error seems to me related to PR with CatalogInventory, that was already merged to phpunit8 branch, so it shouldn't cause this issue anymore.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@slavvka could you get the log from unit tests running now?

**Comment by slavvka** [comment]
```
07:36:21 PHP Fatal error:  Declaration of Magento\ConfigurableProduct\Test\Unit\Ui\DataProvider\Product\Form\Modifier\CustomOptionsTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/Form/Modifier/CustomOptionsTest.php on line 46
```

**Comment by ihor-sviziev** [comment]
@slavvka not is seems to me dependent on https://github.com/magento/magento2/pull/27822.
I'm reviewing it right now

**Comment by slavvka** [comment]
```
02:33:03 PHP Fatal error:  Declaration of Magento\ConfigurableProduct\Test\Unit\Ui\DataProvider\Product\Form\Modifier\CustomOptionsTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/Form/Modifier/CustomOptionsTest.php on line 46
```

**Review by slavvka — CHANGES_REQUESTED** [review]
@lbajsarowicz please fix the unit tests

**Comment by lbajsarowicz** [comment]
@slavvka Fixed. We need to review these tests later on manually to get rid such ridiculous references between modules.

**Review by slavvka — CHANGES_REQUESTED** [review]
@lbajsarowicz 
```
15:42:13 PHP Fatal error:  Declaration of Magento\GiftCard\Test\Unit\Ui\DataProvider\Product\Form\Modifier\GiftCardTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/GiftCard/Test/Unit/Ui/DataProvider/Product/Form/Modifier/GiftCardTest.php on line 247
```

**Comment by slavvka** [comment]
```
11:35:17 There were 13 warnings:
11:35:17 
11:35:17 1) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "missing_previous_and_new_status" (null, null, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 2) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "missing_previous_status_and_subscribe" (null, true, true)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 3) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "new_unsubscribed_value_and_missing_previous_status" (null, false, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 4) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "previous_subscribed_status_without_new_value" (1, null, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 5) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "same_subscribed_previous_and_new_status" (1, true, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 6) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "unsubscribe_previously_subscribed_customer" (1, false, false)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 7) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "previously_unsubscribed_status_without_new_value" (3, null, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 8) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "subscribe_previously_unsubscribed_customer" (3, true, true)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 9) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "same_unsubscribed_previous_and_new_status" (3, false, null)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 10) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "previous_unconfirmed_status_without_new_value" (4, null, true)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 11) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "subscribe_previously_unconfirmed_status" (4, true, true)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 12) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterSave with data set "unsubscribe_previously_unconfirmed_status" (4, false, true)
11:35:17 Trying to configure method "getIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
11:35:17 
11:35:17 13) Magento\Newsletter\Test\Unit\Model\Plugin\CustomerPluginTest::testAfterGetByIdCreatesExtensionAttributes
11:35:17 Trying to configure method "setIsSubscribed" which cannot be configured because it does not exist, has not been specified, is final, or is static
```

**Comment by slavvka** [comment]
```
11:34:50 PHP Fatal error:  Declaration of Magento\GiftCard\Test\Unit\Ui\DataProvider\Product\Form\Modifier\GiftCardTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/GiftCard/Test/Unit/Ui/DataProvider/Product/Form/Modifier/GiftCardTest.php on line 247
```

**Comment by slavvka** [comment]
```
16:05:12 PHP Fatal error:  Declaration of Magento\GiftCard\Test\Unit\Ui\DataProvider\Product\Form\Modifier\GiftCardTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/GiftCard/Test/Unit/Ui/DataProvider/Product/Form/Modifier/GiftCardTest.php on line 247
```

**Comment by slavvka** [comment]
For some reason not related Unit test failed here:
```
12:32:26 There were 2 errors:
12:32:26 
12:32:26 1) Magento\MediaGallery\Test\Unit\Model\Keyword\Command\GetAssetKeywordsTest::testFind with data set "find_one_keyword" (array('keywordRawData'), 1)
12:32:26 Magento\Framework\Exception\IntegrationException: An error occurred during get asset keywords: Warning: Illegal string offset 'id' in /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81.
12:32:26 
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:91
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Test/Unit/Model/Keyword/Command/GetAssetKeywordsTest.php:70
12:32:26 
12:32:26 Caused by
12:32:26 PHPUnit\Framework\Exception: Warning: Illegal string offset 'id' in /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81.
12:32:26 
12:32:26 /var/www/html/dev/tests/unit/framework/bootstrap.php:54
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Test/Unit/Model/Keyword/Command/GetAssetKeywordsTest.php:70
12:32:26 
12:32:26 2) Magento\MediaGallery\Test\Unit\Model\Keyword\Command\GetAssetKeywordsTest::testFind with data set "find_several_keywords" (array('keywordRawData', 'keywordRawData'), 2)
12:32:26 Magento\Framework\Exception\IntegrationException: An error occurred during get asset keywords: Warning: Illegal string offset 'id' in /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81.
12:32:26 
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:91
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Test/Unit/Model/Keyword/Command/GetAssetKeywordsTest.php:70
12:32:26 
12:32:26 Caused by
12:32:26 PHPUnit\Framework\Exception: Warning: Illegal string offset 'id' in /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81.
12:32:26 
12:32:26 /var/www/html/dev/tests/unit/framework/bootstrap.php:54
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Model/Keyword/Command/GetAssetKeywords.php:81
12:32:26 /var/www/html/app/code/Magento/MediaGallery/Test/Unit/Model/Keyword/Command/GetAssetKeywordsTest.php:70
```

**Comment by slavvka** [comment]
And
```
12:32:00 PHP Fatal error:  Declaration of Magento\Banner\Test\Unit\Block\Adminhtml\Widget\Instance\Edit\Chooser\ContainerTest::setUp() must be compatible with Magento\Widget\Test\Unit\Block\Adminhtml\Widget\Instance\Edit\Chooser\AbstractContainerTest::setUp(): void in /var/www/html/app/code/Magento/Banner/Test/Unit/Block/Adminhtml/Widget/Instance/Edit/Chooser/ContainerTest.php on line 15
```

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
Looks like `GiftWrapping` module was never processed:
```
17:32:35 PHP Fatal error:  Declaration of Magento\GiftWrapping\Test\Unit\Ui\DataProvider\Product\Form\Modifier\GiftWrappingTest::setUp() must be compatible with Magento\Catalog\Test\Unit\Ui\DataProvider\Product\Form\Modifier\AbstractModifierTest::setUp(): void in /var/www/html/app/code/Magento/GiftWrapping/Test/Unit/Ui/DataProvider/Product/Form/Modifier/GiftWrappingTest.php on line 106
```

**Comment by slavvka** [comment]
And I don't knwo why MediGallery tests are failing here

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@lbajsarowicz No more Unit test fatals, only MediaGallery. And normal Static test failures

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by slavvka** [comment]
Failed unit tests aren't related to the changes

**Review by slavvka — APPROVED** [review]

**Pull request merged**